### PR TITLE
Add markdownlint script to lint docs markup

### LINKED ISSUE #12426: Improve docs tooling (manifest parser and linter)
There are a lot of ways we can improve docs tooling to make the docs team more efficient and make our docs better.

## Manifest parser

We need a new docs manifest parser/generator! Here's what I'd like to see:

1. ~It should use `toc.json` as the source of truth, with `manifest.json` being built as a totally separate/machine-generated artefact.~
2. ~In turn, that means that `manifest.json` should pull page titles from the markdown, rather than needing to set it in multiple places~
3. ~It should be possible to generate documentation (e.g. the data docs) separately from generating the manifest, and this should mean we don't need to generate an intermediate `root-manifest.json` anymore~
4. `toc.json` should be extended to provide a wildcard option, so auto-generated docs don't need to be explicitly listed inside the file, and can be included in the correct place inside the handbook outline

## Linting

I'd also like to propose the addition of a markdown linter. This is partly to aid the parser, for example to ensure that all docs have an h1 title, but also to improve the consistency of our markdown formatting (consistent heading syntax, line breaks, wrapping, etc.). Right now it varies pretty wildly and having a consistent format would be super valuable.

[markdownlint](https://github.com/DavidAnson/markdownlint) and [remark-link](https://github.com/remarkjs/remark-lint) seem like the leading candidates. Does anyone have an opinion or experience on either of these?

To be clear I don't want to lint grammar/spelling, just the markdown _syntax_.

### PR DESCRIPTION
**Pull request by mkaz**
## Description

Adds a new script `lint-md-docs` that runs markdownlint to lint markdown files for proper syntax. This lint PR differs from the lint implemented in #19716, the other linting was for the source code included within the markdown documents This PR and linter lints the markup syntax of the markdown itself.

I, unfortunately, named the first script too generically and think we should rename to `lint-md-src` and then we can use the `lint-md` command to call both. Thoughts?

A default `.markdownlint.json` config is included, it will require some testing to tune. Here are a couple examples of errors it catches:

- MD004  Unordered list style [Expected: dash; Actual: asterisk]
- MD022  Headings should be surrounded by blank lines 
- MD029  Ordered list item prefix [Expected: 2; Actual: 3;
- MD040  Fenced code blocks should have a language specified

See [list of available rules](https://github.com/DavidAnson/markdownlint/blob/master/doc/Rules.md)

Fixes #12426

## How has this been tested?

To test, apply diff and npm, install

You can test a single markdown file specifying on command-line:
`npm run lint-md:docs  README.md`

Or run across all project markdown files:
`npm run lint-md:docs`

It ignores files specified in `.markdownlintignore`

## Types of changes

- New script `lint-md-docs` in scripts package.

- New dependency on [markdownlint](https://github.com/DavidAnson/markdownlint) and [markdownlint-cli](https://github.com/igorshubovych/markdownlint-cli) node modules.

I chose [markdownlint](https://github.com/DavidAnson/markdownlint) package over [remark-lint](https://github.com/remarkjs/remark-lint) because markdownlint seems a bit more active. Plus it has a nice Visual Studio Code extension which I'll document in developer setup once this is accepted.

### CODE REVIEW

**Comment by gziolo** [comment]
> I, unfortunately, named the first script too generically and think we should rename to lint-md-src and then we can use the lint-md command to call both. Thoughts?

How about you you `lint-md-code` or `lint-md-js`? I guess the latter is more appropriate because it uses ESLint.

You shouldn't be worried too much about the decision made. It's a new command that wasn't published to npm yet. We can cherry-pick this PR during the publish flow or call it a breaking change :) 

It would be great to include the `CHANGELOG` entry with whatever you opt for.

Do you plan to enable this new script in `npm run lint` in Gutenberg and as part of the precommit hook?

**Comment by mkaz** [comment]
@gziolo Good suggestion, I updated the previous lint script to `lint-md-js` 
CHANGELOG and README also updated with new names.

My thought would be to get it in, get people testing and tweak the config and then we can add it to the Continuous Integration checks.

File: packages/scripts/README.md [hunk 1]
```diff
@@ -225,23 +226,47 @@ By default, files located in `build` and `node_modules` folders are ignored.
 
 It uses [npm-package-json-lint](https://www.npmjs.com/package/npm-package-json-lint) with the set of recommended rules defined in [@wordpress/npm-package-json-lint-config](https://www.npmjs.com/package/@wordpress/npm-package-json-lint-config) npm package. You can override default rules with your own as described in [npm-package-json-lint wiki](https://github.com/tclindner/npm-package-json-lint/wiki). Learn more in the [Advanced Usage](#advanced-usage) section.
 
-### `lint-md`
+### `lint-md:docs`
```
Comment (gziolo) [inline packages/scripts/README.md]:
```suggestion
### `lint-md-docs`
```

File: packages/scripts/README.md [hunk 2]
```diff
@@ -225,23 +226,47 @@ By default, files located in `build` and `node_modules` folders are ignored.
...
-* `npm run lint:md` - lints markdown files in the entire project’s directories.
+* `npm run lint-md:docs` - lints markdown files in the entire project’s directories.
+
+By default, files located in `build` and `node_modules` folders are ignored.
+
+#### Advanced information
+
+It uses [markdownlint](https://github.com/DavidAnson/markdownlint) with the [.markdownlint.json](https://github.com/WordPress/gutenberg/blob/master/packages/scripts/config/.markdownlint.json) configuration. This configuration tunes the linting rules to match WordPress standard, you can override with your own config, see [markdownlint-cli](https://github.com/igorshubovych/markdownlint-cli/) for command-line parameters.
+
+### `lint-md:js`
```
Comment (gziolo) [inline packages/scripts/README.md]:
```suggestion
### `lint-md-js`
```

File: packages/scripts/README.md [hunk 3]
```diff
@@ -225,23 +226,47 @@ By default, files located in `build` and `node_modules` folders are ignored.
...
+
+Uses eslint to lint the source included in markdown files to enforce standards for JS code.
```
Comment (gziolo) [inline packages/scripts/README.md]: Nit: let's use the name consistently:
```suggestion
Uses ESLint to lint the source included in markdown files to enforce standards for JS code.
```

File: packages/scripts/scripts/lint-md-docs.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+	hasArgInCLI,
+	hasFileArgInCLI,
+	hasProjectFile,
+} = require( '../utils' );
+
+const args = getArgsFromCLI();
+
+const defaultFilesArgs = hasFileArgInCLI() ? [] : [ '**/*.md' ];
+
+// See: https://eslint.org/docs/user-guide/configuring#using-configuration-files-1.
```
Comment (gziolo) [inline packages/scripts/scripts/lint-md-docs.js]: URLs should point to https://github.com/igorshubovych/markdownlint-cli :)

File: packages/scripts/scripts/lint-md-docs.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+// When a configuration is not provided by the project, use from the default
+// provided with the scripts module. Instruct ESLint to avoid discovering via
+// the `--no-eslintrc` flag, as otherwise it will still merge with inherited.
+const defaultConfigArgs = ! hasLintConfig ?
+	[ '--config', fromConfigRoot( '.markdownlint.json' ) ] :
+	[];
+
+// See: https://eslint.org/docs/user-guide/configuring#ignoring-files-and-directories.
+const hasIgnoredFiles = hasArgInCLI( '--ignore' ) ||
```
Comment (gziolo) [inline packages/scripts/scripts/lint-md-docs.js]: `-i` is another flag which allows providing ignored files:
https://github.com/igorshubovych/markdownlint-cli#ignoring-files

File: packages/scripts/scripts/lint-md-docs.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+const defaultFilesArgs = hasFileArgInCLI() ? [] : [ '**/*.md' ];
+
+// See: https://eslint.org/docs/user-guide/configuring#using-configuration-files-1.
+const hasLintConfig = hasArgInCLI( '-c' ) || hasArgInCLI( '--config' );
+
+// When a configuration is not provided by the project, use from the default
+// provided with the scripts module. Instruct ESLint to avoid discovering via
+// the `--no-eslintrc` flag, as otherwise it will still merge with inherited.
+const defaultConfigArgs = ! hasLintConfig ?
+	[ '--config', fromConfigRoot( '.markdownlint.json' ) ] :
```
Comment (gziolo) [inline packages/scripts/scripts/lint-md-docs.js]: This check should be a bit more detailed, see https://github.com/igorshubovych/markdownlint-cli#configuration:
> CLI argument --config is not mandatory. If it is not provided, markdownlint-cli looks for file .markdownlint.json/.markdownlint.yaml/.markdownlint.yml in current folder, or for file .markdownlintrc in current or all upper folders.

File: packages/scripts/README.md [hunk 7]
```diff
@@ -30,7 +30,8 @@ _Example:_
 		"check-licenses": "wp-scripts check-licenses",
 		"lint:css": "wp-scripts lint-style",
 		"lint:js": "wp-scripts lint-js",
-		"lint:md": "wp-scripts lint-md",
+		"lint-md:docs": "wp-scripts lint-md-docs",
```
Comment (gziolo) [inline packages/scripts/README.md]: How do you feel about using: `lint:md-docs` and `lint:md-js` instead?

The idea here is that with libraries like https://www.npmjs.com/package/npm-run-all you can do the following in the `scripts` section:
```json
{
    "scripts": {
        "lint:js": "wp-scripts lint-js",
        "lint:md-docs": "wp-scripts lint-md-docs",
        "lint:md-js": "wp-scripts lint-md-js",
        "lint": "npm-s lint:*"
    }
}

**Review by gziolo — COMMENTED** [review]
> My thought would be to get it in, get people testing and tweak the config and then we can add it to the Continuous Integration checks.

I'm totally fine with the plan 👍 

I left some comments to address. I still need to test it locally but it all looks very nice 💯

**Comment by ntwb** [comment]
Just want to add a comment about naming here, I mentioned elsewhere that I plan on adding a new package, the package is a config, possibly multiple configs to use [remarklint](https://github.com/remarkjs/remark-lint)

Similar to ESLint, stylelint, though for markdown files

It is cross-compatible so using alongside ESLint and stylelint to lint JS & SCSS/CSS inside fenced code blocks in markdown will continue to work as per the design approach taken in this pull request.

So thinking forward any naming scheme nomenclature if possible to allow for this now would be great and potentially avoid a _major_ release to `@wordpress/scripts` in the future.

So, the below takes this into account for the basic concept
```
"lint:css": "wp-scripts lint-style", // stylelint
"lint:js": "wp-scripts lint-js", // ESLint
"lint:md": "wp-scripts lint-md", // remarklint
```

So the extra uses cases to consider here would be for what to name any "linting" inside markdown files for JS & SCSS/CSS

> How do you feel about using: `lint:md-docs` and `lint:md-js` instead?

Extrapolating the above out:

• `lint:md-js`  to lint using ESLint inside markdown JS code blocks
• `lint:md-style`  to lint using stylelint inside markdown SCSS/CSS code blocks

Are any other considerations to consider?

File: packages/scripts/scripts/lint-md-docs.js (hunk 4 above)
Comment (mkaz) [reply packages/scripts/scripts/lint-md-docs.js]: Copy paste miss.

File: packages/scripts/scripts/lint-md-docs.js (hunk 5 above)
Comment (mkaz) [reply packages/scripts/scripts/lint-md-docs.js]: :+1:

File: packages/scripts/scripts/lint-md-docs.js (hunk 6 above)
Comment (mkaz) [reply packages/scripts/scripts/lint-md-docs.js]: Good catch, added additional checks for project file configs

File: packages/scripts/README.md (hunk 7 above)
Comment (mkaz) [reply packages/scripts/README.md]: Makes sense, I was sure which way to go. Updated to this style :+1:

**Comment by mkaz** [comment]
@ntwb I was looking for your comment elsewhere when I started because I thought you had mentioned linting markdown, but wasn't able to find an issue or PR, so first let me apologize if this is duplicate work and confused any matters.

I think remarklint and markdownlint handle the same aspect of linting of the markup inside a markdown document. This PR implements markdownlint to handle that, which I chose because it seemed to be more active, remark only has a handful commits over the past year. I'm not wedded to either and if remark is more powerful by ability to add linting CSS fenced blocks inside markdown I'm fine with switching.

Thoughts?

File: packages/scripts/README.md [hunk 8]
```diff
@@ -225,23 +226,47 @@ By default, files located in `build` and `node_modules` folders are ignored.
...
-Helps enforce standards for JS source code in your markdown files.
+Uses markdownlint to lint the markup of markdown files to enforce standards.
 
 _Example:_
 
 ```json
 {
 	"scripts": {
-		"lint:md": "wp-scripts lint-md"
+		"lint-md:docs": "wp-scripts lint-md-docs"
```
Comment (gziolo) [inline packages/scripts/README.md]:
```suggestion
		"lint:md-docs": "wp-scripts lint-md-docs"
```

File: packages/scripts/README.md [hunk 9]
```diff
@@ -225,23 +226,47 @@ By default, files located in `build` and `node_modules` folders are ignored.
...
 	}
 }
 ```
 
 This is how you execute the script with presented setup:
 
-* `npm run lint:md` - lints markdown files in the entire project’s directories.
+* `npm run lint-md:docs` - lints markdown files in the entire project’s directories.
```
Comment (gziolo) [inline packages/scripts/README.md]:
```suggestion
* `npm run lint:md-docs` - lints markdown files in the entire project’s directories.
```

File: packages/scripts/README.md [hunk 10]
```diff
@@ -225,23 +226,47 @@ By default, files located in `build` and `node_modules` folders are ignored.
...
+### `lint-md-js`
+
+Uses ESLint to lint the source included in markdown files to enforce standards for JS code.
+
+_Example:_
+
+```json
+{
+	"scripts": {
+		"lint-md:js": "wp-scripts lint-md-js"
```
Comment (gziolo) [inline packages/scripts/README.md]:
```suggestion
		"lint:md-js": "wp-scripts lint-md-js"
```

File: packages/scripts/README.md [hunk 11]
```diff
@@ -225,23 +226,47 @@ By default, files located in `build` and `node_modules` folders are ignored.
...
+	}
+}
+```
+
+This is how you execute the script with presented setup:
+
+* `npm run lint-md:js` - lints markdown files in the entire project’s directories.
```
Comment (gziolo) [inline packages/scripts/README.md]:
```suggestion
* `npm run lint:md-js` - lints markdown files in the entire project’s directories.
```

File: packages/scripts/scripts/lint-md-docs.js [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+	[];
+
+// See: https://github.com/igorshubovych/markdownlint-cli#ignoring-files
+// Check if ignore specified on command-line or project file
+const hasIgnoredFiles = hasArgInCLI( '--ignore' ) || hasArgInCLI( '-i' ) ||
+	hasProjectFile( '.markdownlintignore' );
+
+// Default ignore [ build, node_modules ] directories
+// TODO: Once https://github.com/igorshubovych/markdownlint-cli/issues/46 is in
```
Comment (gziolo) [inline packages/scripts/scripts/lint-md-docs.js]: Great idea to add a comment with future improvements possible.

File: packages/scripts/README.md [hunk 13]
```diff
@@ -30,7 +30,8 @@ _Example:_
 		"check-licenses": "wp-scripts check-licenses",
 		"lint:css": "wp-scripts lint-style",
 		"lint:js": "wp-scripts lint-js",
-		"lint:md": "wp-scripts lint-md",
+		"lint-md:docs": "wp-scripts lint-md-docs",
+		"lint-md:js": "wp-scripts lint-md-js",
```
Comment (gziolo) [inline packages/scripts/README.md]:
```suggestion
		"lint:md:js": "wp-scripts lint-md-js",
```

File: packages/scripts/README.md (hunk 7 above)
Comment (gziolo) [inline packages/scripts/README.md]:
```suggestion
		"lint:md:docs": "wp-scripts lint-md-docs",
```

**Review by gziolo — COMMENTED** [review]
How about we go more granular and reference in examples these new commands as:
- `lint:md:js`
- `lint:md:docs`

It's hard to come up with a good convention :)

**Comment by gziolo** [comment]
I did the following:
```
$ npm --version                                             
6.13.7
$ npm install
```

There were changes in the lock file: 930251a.

File: package.json [hunk 14]
```diff
@@ -194,7 +194,8 @@
 		"lint-css": "wp-scripts lint-style '**/*.scss'",
 		"lint-css:fix": "npm run lint-css -- --fix",
 		"lint-types": "tsc",
-		"lint-md": "wp-scripts lint-md",
+		"lint:md-js": "wp-scripts lint-md-js",
```
Comment (gziolo) [inline package.json]: We could bring some structure to the names of all those scripts in the main `package.json` file as an independent patch 😄

**Comment by gziolo** [comment]
I tested it locally. I must admit it's very noisy even when you lint a subset of files:

```bash
$ npm run lint:md-docs packages/scripts

packages/scripts/CHANGELOG.md:1 MD041/first-line-heading/first-line-h1 First line in file should be a top level heading [Context: "## Master"]
...
packages/scripts/CHANGELOG.md:51 MD024/no-duplicate-heading/no-duplicate-header Multiple headings with the same content [Context: "### Bug Fixes"]
...
packages/scripts/README.md:166 MD004/ul-style Unordered list style [Expected: asterisk; Actual: dash]
...
packages/scripts/README.md:419 MD024/no-duplicate-heading/no-duplicate-header Multiple headings with the same content [Context: "#### Advanced information"]
```

File: packages/scripts/scripts/lint-md-docs.js [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+// we can switch this to specify an ignore file on the command-line. By default,
+// markdownlint looks for .markdownlintignore in project direcotry, but how our
+// scripts work we store the configs in the scripts/config directory
+const defaultIgnoreArgs = ! hasIgnoredFiles ?
+	[ '--ignore', 'build', '--ignore', 'node_modules' ] :
```
Comment (gziolo) [inline packages/scripts/scripts/lint-md-docs.js]: I tested locally with `npm run lint:md-docs packages/scripts` with the markdown lint ignore file removed and it lists files from `node_modules`. The following change fixes it:

```suggestion
	[ '--ignore', '**/build/**', '--ignore', '**/node_modules/**' ] :
```

**Review by gziolo — APPROVED** [review]
Awesome, I left some more comments but only one is important. The one about the default ignores handling. I now realized I haven't tested it with the top folder but it might be difficult with Gutenberg alone.

In my testing, everything works as expected. I could define my own config and override some of the errors reported 👍 

I no longer know what's the best way to name those scripts in the provided `package.json` example.  It's a hard issue to solve 😅

File: package.json (hunk 14 above)
Comment (mkaz) [reply package.json]: Ticket opened: https://github.com/WordPress/gutenberg/issues/19949

File: packages/scripts/scripts/lint-md-docs.js (hunk 12 above)
Comment (mkaz) [reply packages/scripts/scripts/lint-md-docs.js]: Thanks, I'll forget to follow up without it

**Comment by ntwb** [comment]
Thanks @mkaz 

> @ntwb I was looking for your comment elsewhere when I started because I thought you had mentioned linting markdown, but wasn't able to find an issue or PR, so first let me apologize if this is duplicate work and confused any matters.

No apology required, I couldn't find it either 🤣

I've also only been working on this locally, so no code pushed up here.

> I think remarklint and markdownlint handle the same aspect of linting of the markup inside a markdown document. This PR implements markdownlint to handle that, which I chose because it seemed to be more active, remark only has a handful commits over the past year. I'm not wedded to either and if remark is more powerful by having the ability to add linting CSS fenced blocks inside markdown I'm fine with switching.

I also have no strong preference, I think with remarklint being more modulised in that each rule is its own package on npm commit activity makes it harder to determine project activity...

Nevertheless, happy to go with markdownlint, I just has a quick search of the rules I'd be looking at: (where MDxxxx is the equiv rule)

```js
			["lint-emphasis-marker","_"],
			["lint-list-item-indent","space"], MD007
			["lint-maximum-line-length", false], MD013
			["lint-no-html"], MD033 
			["lint-ordered-list-marker-style","."], MD029
			["lint-ordered-list-marker-value","ordered"], MD029
			["lint-strong-marker", "*"],
			["lint-table-pipe-alignment", false]
```

This should all be fine the emphasis-marker and strong-marker rules I'll have to investigate further if there's a markdownlint rule available, though this could have been a personal preference for consistency that I need to double check

----

The background on all of the above is that quite a few projects sync markdown from  GitHub to w.org handbooks, I think there's around 6 repos at the moment.

This is all powered by a custom w.orf wp-cli command that transpiles some of the markdown on the way in before Jetpack it to the transpiles it and finally landing in a handbook page.

As such, there are a handful of idiosyncrasies that require the markdown to be formatted in a specific way so that the end result is it displays correctly in the handbook.

**Review by ntwb — APPROVED** [review]

**Comment by mkaz** [comment]
Thanks for the additional context, I didn't realize there were so many processes, definitely getting a standardized config will help matters. :+1:

File: packages/scripts/scripts/lint-md-docs.js (hunk 12 above)
Comment (youknowriad) [reply packages/scripts/scripts/lint-md-docs.js:41]: I just stumbled on this one, looks like the related issue is now fixed. So should we follow-up?

Comment (mkaz) [reply packages/scripts/scripts/lint-md-docs.js:41]: 👍 Updated in #32633

**Pull request merged**